Add a Windows CI Run

Lets ensure we stay working on Windows.

1) Check all unit tests pass on Windows
2) Add us to appveyor or any other service that gives us Windows containers etc.

I just started to work on this issue during the PyPA Sprint Weekend at Bloomberg (2018). I think there is a dependency that has to be worked on module mirror.py, using fctnl not supported on windows. 
https://ci.appveyor.com/project/charlievil/bandersnatch/builds/19864677/job/tt6qv62c3dc6nlaf#L110

```
__________ ERROR collecting src/bandersnatch/tests/test_mirror.py ____________
ImportError while importing test module 'C:\projects\bandersnatch\src\bandersnatch\tests\test_mirror.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
src\bandersnatch\tests\test_mirror.py:12: in <module>
    from bandersnatch.mirror import Mirror
.tox\py\lib\site-packages\bandersnatch\mirror.py:4: in <module>
    import fcntl
E   ModuleNotFoundError: No module named 'fcntl
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -29,7 +29,7 @@ jobs:
 
     - name: Run Unittests
       run: |
-        export TOXENV=py$(echo -n "${{ matrix.python-version }}" | tr -d '.')
+        export TOXENV=py3
         python test_runner.py
 
     - name: Run Integration Test

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -9,7 +9,7 @@ jobs:
     strategy:
       matrix:
         python-version: [3.6, 3.7, 3.8]
-        os: [macOS-latest, ubuntu-latest]
+        os: [macOS-latest, ubuntu-latest, windows-latest]
 
     steps:
     - uses: actions/checkout@v1

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -28,8 +28,9 @@ jobs:
         python -m pip install -r requirements.txt
 
     - name: Run Unittests
+      env:
+        TOXENV: py3
       run: |
-        export TOXENV=py3
         python test_runner.py
 
     - name: Run Integration Test

--- a/src/bandersnatch/package.py
+++ b/src/bandersnatch/package.py
@@ -91,7 +91,8 @@ class Package:
             self.normalized_legacy_simple_directory,
         ):
             logger.debug(f"Checking if {deprecated_dir} != {self.simple_directory}")
-            if deprecated_dir != self.simple_directory:
+            # Had to compare path strs as Windows did not match path objects ...
+            if str(deprecated_dir) != str(self.simple_directory):
                 if not deprecated_dir.exists():
                     logger.debug(f"{deprecated_dir} does not exist. Not cleaning up")
                     continue

--- a/src/bandersnatch/package.py
+++ b/src/bandersnatch/package.py
@@ -85,12 +85,15 @@ class Package:
         if not self.cleanup:
             return
 
+        logger.debug(f"Running Non PEP503 path cleanup for {self.raw_name}")
         for deprecated_dir in (
             self.raw_simple_directory,
             self.normalized_legacy_simple_directory,
         ):
+            logger.debug(f"Checking if {deprecated_dir} != {self.simple_directory}")
             if deprecated_dir != self.simple_directory:
                 if not deprecated_dir.exists():
+                    logger.debug(f"{deprecated_dir} does not exist. Not cleaning up")
                     continue
 
                 logger.info(

--- a/src/bandersnatch/tests/test_mirror.py
+++ b/src/bandersnatch/tests/test_mirror.py
@@ -249,7 +249,9 @@ web{0}packages{0}any{0}f{0}foo{0}foo.zip
 web{0}simple
 web{0}simple{0}foobar
 web{0}simple{0}foobar{0}index.html
-web{0}simple{0}index.html""".format(sep)
+web{0}simple{0}index.html""".format(
+        sep
+    )
     if WINDOWS:
         expected = expected.replace(".lock\n", "")
     assert expected == utils.find(mirror.homedir)
@@ -319,12 +321,12 @@ todo
 web{0}packages{0}2.7{0}f{0}foo{0}foo.whl
 web{0}packages{0}any{0}f{0}foo{0}foo.zip
 web{0}simple{0}foo{0}index.html
-web{0}simple{0}index.html""".format(sep)
+web{0}simple{0}index.html""".format(
+        sep
+    )
     if WINDOWS:
         expected = expected.replace(".lock\n", "")
-    assert expected == utils.find(
-        mirror.homedir, dirs=False
-    )
+    assert expected == utils.find(mirror.homedir, dirs=False)
     assert (
         open("web{0}simple{0}index.html".format(sep)).read()
         == """\

--- a/src/bandersnatch/tests/test_mirror.py
+++ b/src/bandersnatch/tests/test_mirror.py
@@ -11,6 +11,7 @@ from bandersnatch import utils
 from bandersnatch.configuration import BandersnatchConfig, Singleton
 from bandersnatch.filter import filter_project_plugins
 from bandersnatch.mirror import Mirror
+from bandersnatch.utils import WINDOWS
 
 
 class JsonDict(dict):
@@ -228,7 +229,7 @@ async def test_mirror_empty_resume_from_todo_list(mirror):
 
     await mirror.synchronize()
 
-    assert """\
+    expected = """\
 .lock
 generation
 status
@@ -248,11 +249,11 @@ web{0}packages{0}any{0}f{0}foo{0}foo.zip
 web{0}simple
 web{0}simple{0}foobar
 web{0}simple{0}foobar{0}index.html
-web{0}simple{0}index.html""".format(
-        sep
-    ) == utils.find(
-        mirror.homedir
-    )
+web{0}simple{0}index.html""".format(sep)
+    if WINDOWS:
+        expected = expected.replace(".lock\n", "")
+    assert expected == utils.find(mirror.homedir)
+
     assert (
         open("web{0}simple{0}index.html".format(sep)).read()
         == """\
@@ -311,16 +312,17 @@ async def test_mirror_sync_package_error_no_early_exit(mirror):
     mirror.errors = True
     changed_packages = await mirror.synchronize()
 
-    assert """\
+    expected = """\
 .lock
 generation
 todo
 web{0}packages{0}2.7{0}f{0}foo{0}foo.whl
 web{0}packages{0}any{0}f{0}foo{0}foo.zip
 web{0}simple{0}foo{0}index.html
-web{0}simple{0}index.html""".format(
-        sep
-    ) == utils.find(
+web{0}simple{0}index.html""".format(sep)
+    if WINDOWS:
+        expected = expected.replace(".lock\n", "")
+    assert expected == utils.find(
         mirror.homedir, dirs=False
     )
     assert (

--- a/src/bandersnatch/tests/test_package.py
+++ b/src/bandersnatch/tests/test_package.py
@@ -33,7 +33,7 @@ async def test_cleanup_non_pep_503_paths(mirror):
     await package.cleanup_non_pep_503_paths()
 
     # Create a non normailized directory
-    touch_files([Path(f"web/simple/{raw_package_name}/index.html")])
+    touch_files([package.mirror.webdir / "simple" / raw_package_name / "index.html"])
 
     package.cleanup = True
     with mock.patch("bandersnatch.package.rmtree") as mocked_rmtree:

--- a/src/bandersnatch/tests/test_utils.py
+++ b/src/bandersnatch/tests/test_utils.py
@@ -2,7 +2,7 @@ import os
 import os.path
 import re
 from pathlib import Path
-from tempfile import TemporaryDirectory, gettempdir
+from tempfile import NamedTemporaryFile, TemporaryDirectory, gettempdir
 
 import aiohttp
 import pytest
@@ -31,14 +31,21 @@ def test_convert_url_to_path():
 
 
 def test_hash():
-    sample = os.path.join(os.path.dirname(__file__), "sample")
-    assert hash(sample, function="md5") == "125765989403df246cecb48fa3e87ff8"
-    assert hash(sample, function="sha256") == (
-        "95c07c174663ebff531eed59b326ebb3fa95f418f680349fc33b07dfbcf29f18"
-    )
-    assert hash(sample) == (
-        "95c07c174663ebff531eed59b326ebb3fa95f418f680349fc33b07dfbcf29f18"
-    )
+    expected_md5 = "b2855c4a4340dad73d9d870630390885"
+    expected_sha256 = "a2a5e3823bf4cccfaad4e2f0fbabe72bc8c3cf78bc51eb396b5c7af99e17f07a"
+    with NamedTemporaryFile(delete=False) as ntf:
+        ntf_path = Path(ntf.name)
+        ntf.close()
+        try:
+            with ntf_path.open("w") as ntfp:
+                ntfp.write("Unittest File for hashing Fun!")
+
+            assert hash(ntf_path, function="md5") == expected_md5
+            assert hash(ntf_path, function="sha256") == expected_sha256
+            assert hash(ntf_path) == expected_sha256
+        finally:
+            if ntf_path.exists():
+                ntf_path.unlink()
 
 
 def test_find_files():

--- a/test_runner.py
+++ b/test_runner.py
@@ -19,7 +19,9 @@ from tempfile import gettempdir
 
 from src.bandersnatch.utils import hash
 
-BANDERSNATCH_EXE = Path(which("bandersnatch") or "bandersnatch")
+BANDERSNATCH_EXE = Path(
+    which("bandersnatch") or which("bandersnatch.exe") or "bandersnatch"
+)
 CI_CONFIG = Path("src/bandersnatch/tests/ci.conf")
 EOP = "[CI ERROR]:"
 MIRROR_ROOT = Path(f"{gettempdir()}/pypi")
@@ -114,7 +116,6 @@ def do_ci(conf: Path) -> int:
 
 def platform_config() -> Path:
     """Ensure the CI_CONFIG is correct for the platform we're running on"""
-    global BANDERSNATCH_EXE
     platform_ci_conf = MIRROR_ROOT / "ci.conf"
     cp = ConfigParser()
     cp.read(str(CI_CONFIG))
@@ -124,10 +125,6 @@ def platform_config() -> Path:
 
     with platform_ci_conf.open("w") as pccfp:
         cp.write(pccfp)
-
-    # TODO: Correct path + see if this actually works
-    if system() == "Windows":
-        BANDERSNATCH_EXE = Path(r"C:\pip\bin\bandersnatch.exe")
 
     return platform_ci_conf
 

--- a/test_runner.py
+++ b/test_runner.py
@@ -11,7 +11,6 @@ then check for expected outputs to exist
 from configparser import ConfigParser
 from os import environ
 from pathlib import Path
-from platform import system
 from shutil import rmtree, which
 from subprocess import run
 from sys import exit

